File: reference/two-pointer/1498-Number-of-Subsequences-That-Satisfy-the-Given-Sum-Condition.go
func numSubseq(nums []int, target int) int {
    const mod = 1000000007
    ans := 0
    sort.Ints(nums)
    left, right := 0, len(nums) - 1
    for left <= right {
        if nums[left] + nums[right] <= target {
            ans += powMod(2, right - left, mod)
            ans %= mod
            left++
        } else {
            right--
        }
    }
    return ans
}

func powMod(base, exp, mod int) int {
    result := 1
    base = base % mod
    for exp > 0 {
        if exp % 2 == 1 {
            result = (result * base) % mod
        }
        exp = exp >> 1
        base = (base * base) % mod
    }
    return result
}

File: reference/two-pointer/1498-Number-of-Subsequences-That-Satisfy-the-Given-Sum-Condition.py
class Solution(object):
    def numSubseq(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        MOD = 10**9 + 7
        ans = 0
        left, right = 0, len(nums) - 1
        while left <= right:
            if nums[left] + nums[right] <= target:
                ans += pow(2, right - left)
                ans %= MOD
                left += 1
            else:
                right -= 1
        return ans
